apps/api/src/domain/models/song/commands/publish-song.command-handler.ts

```diff
@@ -1,68 +1,55 @@
-import { ResourceType } from '../../../types/ResourceType';
-import InvalidCommandPayloadTypeError from '../../shared/common-command-errors/InvalidCommandPayloadTypeError';
+import { ResultOrError } from '../../../../types/ResultOrError';
+import { Valid } from '../../../domainModelValidators/Valid';
+import { InMemorySnapshot, ResourceType } from '../../../types/ResourceType';
+import buildInMemorySnapshot from '../../../utilities/buildInMemorySnapshot';
+import { CommandHandlerBase } from '../../shared/command-handlers/command-handler-base';
 import { Song } from '../song.entity';
 import { PublishSong } from './publish-song.command';
 import { SongPublished } from './song-published.event';
 
 @CommandHandler(PublishSong)
-export class PublishSongCommandHandler implements ICommandHandler {
-    constructor(
-        private readonly repositoryProvider: RepositoryProvider,
-        @Inject('ID_MANAGER') private readonly idGenerator: IIdGenerator
-    ) {}
+export class PublishSongCommandHandler extends CommandHandlerBase<Song> {
+    async createOrFetchWriteContext({ id }: PublishSong): Promise<ResultOrError<Song>> {
+        const existingSongSearchResult = await this.repositoryProvider
+            .forResource<Song>(ResourceType.song)
+            .fetchById(id);
 
-    async execute(command: PublishSong): Promise<Ack | InternalError> {
-        const payloadTypeErrors = buildSimpleValidationFunction(
-            Object.getPrototypeOf(command).constructor
-        )(command).map(
-            (simpleError) => new InternalError(`invalid payload type: ${simpleError.toString()}`)
-        );
+        if (isNotFound(existingSongSearchResult))
+            return new InternalError(`There is no song with the id: ${id}`);
 
-        if (payloadTypeErrors.length > 0) {
-            return new InvalidCommandPayloadTypeError('PUBLISH_SONG', payloadTypeErrors);
+        if (isInternalError(existingSongSearchResult)) {
+            throw existingSongSearchResult;
         }
 
-        // validate external state
-        const { id } = command;
-
-        const searchResult = await this.repositoryProvider
-            .forResource<Song>(ResourceType.song)
-            .fetchById(id);
+        return existingSongSearchResult;
+    }
 
-        if (isInternalError(searchResult)) {
-            throw new InternalError(`Encountered an error when fetching song: ${id}`, [
-                searchResult,
-            ]);
-        }
+    actOnInstance(song: Song): ResultOrError<Song> {
+        return song.publish();
+    }
 
-        if (isNotFound(searchResult)) {
-            return new InternalError(`There is no song with id: ${id}`);
-        }
+    async fetchRequiredExternalState(): Promise<InMemorySnapshot> {
+        return buildInMemorySnapshot({});
+    }
 
-        const updatedInstance = searchResult.publish();
+    validateExternalState(_: InMemorySnapshot, __: Song): Valid | InternalError {
```

Comment:
If we end up writing this on multiple handlers, we may want to make this the base implementation on the parent class.